internal/controller/apiextensions/claim/api.go

```diff
@@ -64,7 +64,11 @@ func (a *APIBinder) Bind(ctx context.Context, cm resource.CompositeClaim, cp res
 	}
 
 	// Propagate the actual external name back from the composite to the claim.
-	meta.SetExternalName(cm, meta.GetExternalName(cp))
+	if en, enExists := cp.GetAnnotations()[meta.AnnotationKeyExternalName]; enExists { // we may also add this as a utility to crossplane-runtime
+		meta.SetExternalName(cm, en)
+	} else {
+		meta.RemoveAnnotations(cm, meta.AnnotationKeyExternalName)
+	}
```

Comment:
```suggestion
	if meta.GetExternalName(cm) == "" && meta.GetExternalName(cp) != "" {
		meta.SetExternalName(cm, en)
	}
```

@negz can confirm but I believe the case mentioned here is actually just a late-initialization, i.e. copy the annotation from composite to claim if claim doesn't have a value for it. The bug is that it doesn't respect this order of propagation and overwrites claim's external name no matter what. We'd like to do that only if the claim doesn't have a value for it and composite does.

I'd also suggest using `meta.GetExternalName` because accounting for empty/nil string difference is useful in cases where empty string is actually a valid configuration. In this case, I'm not sure if it's a valid external name in any context. If I remember correctly, that's why we didn't implement it as `meta.GetExternalName(runtime.Object) *string`.

Reply:
Right - iirc that's the goal. We want to propagate the external name back from XR -> XRC _if_ there isn't an external name on the claim already. If there was an external name on the XRC we'd instead propagate it to the XR.